source/Launchpad-Commands/LaunchpadTestProcessingContext.class.st

```diff
@@ -0,0 +1,36 @@
+"
+I provide a convenient way to test Launchpad command handling.
+
+e.g.
+
+	commandLine := CommandLineArguments withArguments: #('start' 'MyApp' '--log-requests=true').
+	testContext := LaunchpadTestProcessingContext handling: commandLine.
+   LaunchpadCommandLineHandler activateInContext: testContext.
+
+	testContext assertRunWithArguments: { true } forTest: self
+"
+Class {
+	#name : #LaunchpadTestProcessingContext,
+	#superclass : #LaunchpadCommandLineProcessingContext,
+	#instVars : [
+		'runArguments'
+	],
+	#category : #'Launchpad-Commands'
+}
+
+{ #category : #'instance creation' }
+LaunchpadTestProcessingContext class >> handling: aCommandLine [
+
+	^ self handling: aCommandLine  writingTo: (WriteStream on: '')
+]
+
+{ #category : #'testing - support' }
+LaunchpadTestProcessingContext >> assertRunWithArguments: aCollection forTest: aTest [
+
+	aTest assertCollection: aCollection hasSameElements: runArguments
+]
+
+{ #category : #'testing - support' }
+LaunchpadTestProcessingContext >> run: aBlock withArguments: aCollection [
+	runArguments := aCollection 
+]
```

Comment:
@gcotelli this is the crux of the problem - what is the hook we can introduce into launchpad where you can check what parameters were extracted and applied to the application you want to run?

The context is the object that stands out for this (and it seems responsible for similar actions like exitFailure etc.

But #basicStartWithin: isn't dispatched via the context - its called by #startWithin: and that begins life via LaunchpadStartApplicationCommand - and this command doesn't have any hook to play with either (e.g. its not retrieved from a context ).

Hence why've added an somewhat artificial run:withArguments: method to allow some form of testing. If you look at the tests I've written - where I pass in a context - you can then effectively assert that method was called.
```
	commandLine := CommandLineArguments withArguments: #('start' 'HTTPServerDemo').
	testContext := LaunchpadTestProcessingContext handling: commandLine.
   LaunchpadCommandLineHandler activateInContext: testContext.

	testContext assertRunWithArguments: { false. 8181} forTest: self
```
I agree in a later comment, that asserting order on those args is a bit naiive - but given its simply an adapter from a config to a live app with an api, I didn't think it was so bad - and the concept could be extended either to expect a dictionary or a block.. but this could all come later.

However, is there another way we can cleanly do this - and get a level of indirection for testing?

Reply:
I think there are two things here.

 If you want to test that the configuration contains the expected values there's already a way to do that.. just start the application, ask it for `configuration`, and assert anything you want over it. 
Now, I think that it's useful to perform these assertions without actually running the application startup code (something like a dry run). In this case, we can change the `startWithin:` code to send a message to the context before running the subclass responsibility message. In summary instead of always sending `self basicStartWithin: context`  do something like `context run: [ self basicStartWithin: context ]`. At this point, all the configuration is already parsed and available.

If you want to test that the configuration values end up in the right place of the application, I think the best option is to start the application and assert that the configuration is used as expected.

So maybe, adding also a general `--dry-run` option to the start command that parses all the configuration but doesn't start the application is a good idea and can be used in a deployed app to check that the configuration is right without starting the app.

Reply:
I committed the more minor changes before reading this (and my commit, didn't address this bit) - but yet perhaps it is a dry-run check that I'm after - something that you can easily put in a test to keep running to get your LaunchpadAppliction easily configured correctly (as the current examples are really clumsy - or at least too low level for someone just wanting to check their configuration is correct).

I still think there is something interesting in how a real life application might "adapt" the configuration sent from launchpad - although maybe thats just a case of going all in on launchpad configuration being the configuration you pass into your app? In my my case - as I didn't start this way - I could write a separate adapter over the launchpad one and just test that in isolation. If that was the case my HTTPServerDemo basicStartIn: would probably look something like this ?

```
LaunchpadHTTPServerDemoApplication startWithConfig: self configuration
```

But is it a good idea to tie yourself to LaunchPad so directly - that makes me feel nervous...  perhaps it might be useful if ApplicationConfiguration also had an #asDictionary option - so an app can just look keys?

I still think we should consider what a dry-run test would look like though?

'''
        commandLine := CommandLineArguments withArguments: #( 'launchpad' 'start' '-dryrun' 'HTTPServerDemo' '--port=8080' '--log-requests=true' ).

	config := LaunchpadCommandLineHandler activateWith: commandLine.

        self assert: config port equals: 8080.
       etc.
```

That would be a clean test - you can dryrun without an exception, and you can assert some of the config items if you are doing something tricky in them?